dataframely/_storage/delta.py

```diff
@@ -0,0 +1,201 @@
+        import deltalake
+
+        uri = Path(kwargs.pop("source"))
+
+        data = {}
+        collection_types = []
+        for key in members:
+            member_uri = uri / key
+            if not deltalake.DeltaTable.is_deltatable(str(member_uri)):
+                continue
+            table = _to_delta_table(member_uri)
+            data[key] = pl.scan_delta(table, **kwargs)
+            collection_types.append(_read_serialized_collection(table))
+
+        return data, collection_types
+
+    def read_collection(
+        self, members: Iterable[str], **kwargs: Any
+    ) -> tuple[dict[str, pl.LazyFrame], list[SerializedCollection | None]]:
+        lazy, collection_types = self.scan_collection(members, **kwargs)
+        eager = {name: lf.collect().lazy() for name, lf in lazy.items()}
+        return eager, collection_types
+
+    # ------------------------------ Failure Info --------------------------------------
+    def sink_failure_info(
+        self,
+        lf: pl.LazyFrame,
+        serialized_rules: SerializedRules,
+        serialized_schema: SerializedSchema,
+        **kwargs: Any,
+    ) -> None:
+        _raise_on_lazy_write()
+
+    def write_failure_info(
+        self,
+        df: pl.DataFrame,
+        serialized_rules: SerializedRules,
+        serialized_schema: SerializedSchema,
+        **kwargs: Any,
+    ) -> None:
+        self.write_frame(
+            df,
+            serialized_schema,
+            metadata={
+                RULE_METADATA_KEY: serialized_rules,
+            },
+            **kwargs,
+        )
+
+    def scan_failure_info(
+        self, **kwargs: Any
+    ) -> tuple[pl.LazyFrame, SerializedRules, SerializedSchema]:
+        """Lazily read the failure info from the storage backend."""
+        table = _to_delta_table(kwargs.pop("source"))
+
+        # Metadata
+        serialized_rules = assert_metadata(_read_serialized_rules(table))
+        serialized_schema = assert_metadata(_read_serialized_schema(table))
+
+        # Data
+        lf = pl.scan_delta(table, **kwargs)
+
+        return lf, serialized_rules, serialized_schema
+
+
+def _raise_on_lazy_write() -> None:
+    raise NotImplementedError("Lazy writes are not currently supported for deltalake.")
+
+
+def _read_serialized_schema(table: "deltalake.DeltaTable") -> SerializedSchema | None:
+    [last_commit] = table.history(limit=1)
+    return last_commit.get(SCHEMA_METADATA_KEY, None)
+
+
+def _read_serialized_collection(
+    table: "deltalake.DeltaTable",
+) -> SerializedCollection | None:
+    [last_commit] = table.history(limit=1)
+    return last_commit.get(COLLECTION_METADATA_KEY, None)
+
+
+def _read_serialized_rules(
+    table: "deltalake.DeltaTable",
+) -> SerializedRules | None:
+    [last_commit] = table.history(limit=1)
+    return last_commit.get(RULE_METADATA_KEY, None)
+
+
+def _to_delta_table(
+    table: "Path | str | deltalake.DeltaTable",
+) -> "deltalake.DeltaTable":
```

Comment:
You can just use `deltalake.DeltaTable` in the annotations (i.e. without quotes) here if you use `from __future__ import annotations`

Reply:
ahh so that's why I only needed the quotes sometimes and not others. 👍